src/rust/engine/client/src/client.rs

```diff
@@ -0,0 +1,128 @@
+use tokio::io::AsyncWriteExt;
+use tokio::net::TcpStream;
+
+pub struct ConnectionSettings {
+  pub port: u16,
+  pub timeout_limit: f64,
+  pub dynamic_ui: bool,
+}
+
+impl ConnectionSettings {
+  pub fn default(port: u16) -> ConnectionSettings {
+    ConnectionSettings {
+      port,
+      timeout_limit: 60.0,
+      dynamic_ui: true,
+    }
+  }
+}
+
+pub async fn execute_command(
+  start: SystemTime,
+  connection_settings: ConnectionSettings,
+  mut env: Vec<(String, String)>,
+  argv: Vec<String>,
+  working_dir: &PathBuf,
+) -> Result<i32, String> {
+  env.push((
+    "PANTSD_RUNTRACKER_CLIENT_START_TIME".to_owned(),
+    start
+      .duration_since(SystemTime::UNIX_EPOCH)
+      .map_err(|e| format!("Failed to determine current unix time: {err}", err = e))?
+      .as_secs_f64()
+      .to_string(),
+  ));
+  env.push((
+    "PANTSD_REQUEST_TIMEOUT_LIMIT".to_owned(),
+    connection_settings.timeout_limit.to_string(),
+  ));
+  if connection_settings.dynamic_ui {
+    for raw_fd in &[
+      std::io::stdin().as_raw_fd(),
+      std::io::stdout().as_raw_fd(),
+      std::io::stderr().as_raw_fd(),
+    ] {
+      if let Ok(path) = nix::unistd::ttyname(*raw_fd) {
+        env.push((
+          format!("NAILGUN_TTY_PATH_{fd}", fd = raw_fd),
+          path.display().to_string(),
+        ));
+      }
+    }
+  }
+
+  let cmd = Command {
+    command: argv
+      .get(0)
+      .ok_or_else(|| "Failed to determine current process argv0".to_owned())?
+      .clone(),
+    args: argv.iter().skip(1).cloned().collect(),
+    env,
+    working_dir: working_dir.into(),
+  };
+
+  // TODO: This aligns with the C client. Possible that the default client and server configs
+  // should be different in order to be maximally lenient.
+  let config = nails::Config::default().heartbeat_frequency(Duration::from_millis(500));
+
+  debug!(
+    "Connecting to server at {address:?}...",
+    address = &connection_settings.port
+  );
+  let stream = TcpStream::connect(("0.0.0.0", connection_settings.port))
+    .await
+    .map_err(|e| format!("Error connecting to pantsd: {err}", err = e))?;
+  let mut child = nails::client::handle_connection(config, stream, cmd, async {
+    let (stdin_write, stdin_read) = child_channel::<ChildInput>();
+    let _join = tokio::spawn(handle_stdin(stdin_write));
+    stdin_read
+  })
+  .map_err(|e| format!("Error starting process: {err}", err = e))
+  .await?;
+
+  let output_stream = child.output_stream.take().unwrap();
+  let stdio_printer = async move { tokio::spawn(handle_stdio(output_stream)).await.unwrap() };
+
+  future::try_join(stdio_printer, child.wait())
+    .await
+    .map(|(_, exit_code)| exit_code.0)
+    .map_err(|e| format!("Error executing process: {err}", err = e))
+}
+
+async fn handle_stdio(
+  mut stdio_read: impl Stream<Item = ChildOutput> + Unpin,
+) -> Result<(), io::Error> {
+  let mut stdout = tokio::io::stdout();
```

Comment:
Not sure. I aped this from the nails client. Looking....

Reply:
The tokio variants delegate to the std::io variants without re-configuration. To add buffering of - say - stderr, which is unbuffered IIUC - we could wrap that in a tokio::io::BufWriter. I'm not sure we want to do that though since stderr is traditionally unbuffered. That said I also really don't know what I'm talking about here with terminals and etc in play.